zerver/tornado/handlers.py

```diff
@@ -180,175 +133,112 @@ def delete(self, *args: Any, **kwargs: Any) -> None:
-                response = middleware_method(request, response)
-            if hasattr(self, 'apply_response_fixes'):
-                response = self.apply_response_fixes(request, response)
-        except Exception:  # Any exception should be gathered and handled
-            signals.got_request_exception.send(sender=self.__class__, request=request)
-            response = self.handle_uncaught_exception(request, resolver, sys.exc_info())
-
-        return response
-
-class AsyncDjangoHandler(AsyncDjangoHandlerBase):
-    def zulip_finish(self, response: Dict[str, Any], request: HttpRequest,
+    def zulip_finish(self, result_dict: Dict[str, Any], request: HttpRequest,
                      apply_markdown: bool) -> None:
-        # Make sure that Markdown rendering really happened, if requested.
-        # This is a security issue because it's where we escape HTML.
-        # c.f. ticket #64
-        #
-        # apply_markdown=True is the fail-safe default.
-        if response['result'] == 'success' and 'messages' in response and apply_markdown:
-            for msg in response['messages']:
+        # Function called when we want to break a long-polled
+        # get_events request and return a response to the client.
+        if result_dict['result'] == 'success' and 'messages' in result_dict and apply_markdown:
+            for msg in result_dict['messages']:
                 if msg['content_type'] != 'text/html':
                     self.set_status(500)
                     self.finish('Internal error: bad message format')
-        if response['result'] == 'error':
+        if result_dict['result'] == 'error':
             self.set_status(400)
 
-        # Call the Django response middleware on our object so that
-        # e.g. our own logging code can run; but don't actually use
-        # the headers from that since sending those to Tornado seems
-        # tricky; instead just send the (already json-rendered)
-        # content on to Tornado
-        django_response = json_response(res_type=response['result'],
-                                        data=response, status=self.get_status())
-        django_response = self.apply_response_middleware(request, django_response,
-                                                         request._resolver)
+        # The `result` dictionary contains the data we want to return
+        # to the client.  We want to do so in a proper HTTP response
+        # after running our Django response middleware (which does
+        # things like log the request, add rate-limit headers, etc.).
+        # Making an HttpResponse object with json_response is easy.
+        # What's tricky is getting Django to run process_response
+        # middleware on it without writing a lot of duplicate code.
+        #
+        # Django runs the process_response middleware functions inside
+        # MiddlewareMixin.__call__, which also runs the
+        # process_request middleware.  That code path is activated
```

Comment:
"the process_request middleware" isn't right -- rather that's a method found on middlewares. Same for "to run process_response middleware".

Maybe "... getting Django to run the middlewares' `process_response` methods ..." And then for starting the second paragraph, something like

> Django runs each middleware class's methods inside MiddlewareMixin.__call__, which is invoked inside Django BaseHandler's ..., which together effectively end up doing:

    for m in middlewares:
        m.process_request(request)
    response = self._get_response(request)
    for m in reversed(middlewares):
        response = m.process_response(request, response)

> where the __call__ of each middleware contributes one iteration of each loop.

Provided, of course, that that's true! :stuck_out_tongue: Please fact-check me on that.

(AIUI from just a little reading here, that's true modulo exceptions, the optional short-circuiting behavior of `process_request`, and probably some other API wrinkles that I'm hoping we don't need to worry about here.)

Reply:
In particular I think something like that would help make clear why:
 * we need to knock out calling `process_request` in `MiddlewareMixin.__call__`
 * the handler's `_get_response` is the other spot we need to hook into to short-circuit things